db/Pos/DB/Iterator/DBIterator.hs

```diff
@@ -90,42 +86,53 @@ runIterator dbIter DB {..} =
   where
     run it = do
         Rocks.iterSeek it (iterKeyPrefix @i Proxy)
-        runReaderT (getDBIterator dbIter) it
+        Ether.E.runReaderT Proxy dbIter it
 
 ----------------------------------------------------------------------------
 -- DBMapIterator
 ----------------------------------------------------------------------------
 
+data DBMapIteratorTag i v
+
 -- | Encapsulate `map f elements`, where @elements@ - collection elements of type @a@.
 -- Holds `DBIterator m a` and apply f for every `nextItem` and `curItem` call.
 -- If f :: a -> b then we iterate by collection elements of type b.
-newtype DBMapIterator i v m a = DBMapIterator
-    { getDBMapIterator :: ReaderT (IterType i -> v) (DBIterator i m) a
-    } deriving (Functor, Applicative, Monad, MonadIO, MonadThrow, MonadCatch)
-
-type DBKeyIterator   i = DBMapIterator i (IterKey i)
-type DBValueIterator i = DBMapIterator i (IterValue i)
-
-instance MonadTrans (DBMapIterator i v)  where
-    lift x = DBMapIterator $ ReaderT $ const $ lift x
+type DBMapIterator i v =
+  Ether.E.ReaderT (DBMapIteratorTag i v) (IterType i -> v, Rocks.Iterator)
+
+dbMapIterator
+  :: ((IterType i -> v) -> DBIterator i m a)
+  -> DBMapIterator i v m a
+dbMapIterator mkDbIter =
+  Ether.E.readerT Proxy $ \(f, ri) ->
+    Ether.E.runReaderT Proxy (mkDbIter f) ri
+
+unDbMapIterator
+  :: (IterType i -> v)
+  -> DBMapIterator i v m a
+  -> DBIterator i m a
+unDbMapIterator f dmi =
+  Ether.E.readerT Proxy $ \ri ->
+    Ether.E.runReaderT Proxy dmi (f, ri)
+
+type DBKeyIterator   i v = DBMapIterator i (IterKey i) v
+type DBValueIterator i v = DBMapIterator i (IterValue i) v
 
 -- -- | Instance for DBMapIterator using DBIterator.
 -- -- Fetch every element from DBIterator and apply `f` for it.
 instance ( Monad m
-         , MonadIterator (IterKey i, IterValue i) (DBIterator i m))
-         => MonadIterator v (DBMapIterator i v m) where
-    nextItem = DBMapIterator $ ReaderT $ \f -> fmap f <$> nextItem
-    curItem = DBMapIterator $ ReaderT $ \f -> fmap f <$> curItem
-
-deriving instance MonadMask m => MonadMask (DBMapIterator i v m)
-deriving instance MonadDB m => MonadDB (DBMapIterator i v m)
+         , MonadIterator (IterKey i, IterValue i) (DBIterator i m)
+         , p ~ (IterType i -> v, Rocks.Iterator))
```

Comment:
Throughout this request type equality in instances is used mostly according to the pattern described in http://chrisdone.com/posts/haskell-constraint-trick

The point is to make GHC commit to an instance without inspecting some parts of the type.